osu.Game/Screens/Edit/Screens/Setup/Components/OsuColourPickerGradient.cs

```diff
@@ -0,0 +1,168 @@
+        public const float SIZE_Y = 180;
+        public const float COLOUR_LABEL_TEXT_SIZE = 18;
+
+        public event Action<Color4> SelectedColourChanged;
+
+        public void TriggerSelectedColourChanged(Color4 newValue)
+        {
+            SelectedColourChanged?.Invoke(newValue);
+        }
+
+        private Color4 activeColour;
+        public Color4 ActiveColour
+        {
+            get => activeColour;
+            set
+            {
+                if (value == activeColour)
+                    return;
+
+                activeColour = value;
+                pickerContainer.Clear();
+                loadPicker();
+                isColourChangedFromActiveColour = true;
+                Current.Value = calculateColour(selectedColourPosition);
+                isColourChangedFromActiveColour = false;
+            }
+        }
+
+        public float ActiveHue => Color4.ToHsv(activeColour).X;
+
+        public OsuColourPickerGradient()
+        {
+            CircularContainer colourPickerButton;
+
+            Size = new Vector2(SIZE_X, SIZE_Y);
+
+            Children = new Drawable[]
+            {
+                pickerContainer = new Container
+                {
+                    Size = new Vector2(SIZE_X, SIZE_Y),
+                    CornerRadius = 10,
+                    Masking = true,
+                },
+                colourPickerButton = new CircularContainer
+                {
+                    AlwaysPresent = true,
+                    RelativeSizeAxes = Axes.None,
+                    RelativePositionAxes = Axes.None,
+                    Origin = Anchor.Centre,
+                    Anchor = Anchor.TopLeft,
+                    Size = new Vector2(15),
+                    //Position = new Vector2(50),
+                    BorderThickness = 2,
+                    BorderColour = Color4.White,
+                }
+            };
+
+            loadPicker();
+
+            Current.ValueChanged += newValue =>
+            {
+                if (!isColourChangedFromActiveColour)
+                {
+                    if (!isColourChangedFromGradient)
+                    {
+                        var hsv = Color4.ToHsv(newValue);
+                        selectedColourPosition.X = hsv.Y * SIZE_X;
+                        selectedColourPosition.Y = (1 - hsv.Z) * SIZE_Y;
+                    }
+                    colourPickerButton.Position = selectedColourPosition;
+                }
+                colourPickerButton.Colour = newValue;
+                TriggerSelectedColourChanged(newValue);
+            };
+        }
+
+        private void loadPicker()
+        {
+            pickerContainer.Add(new Box
+            {
+                RelativeSizeAxes = Axes.Both,
+                Colour = Color4.Black,
+            });
+            Color4 transparentBlack = new Color4(0, 0, 0, 0);
+            for (float x = 0; x < SIZE_X; x++)
+            {
+                float white = (SIZE_X - x) / SIZE_X;
```

Comment:
Since `x` will always be `< SIZE_X`, you can never choose 100% saturation.

Reply:
It will just never be drawn, in reality, the user may choose any colour. The reason why it's implemented like that is because the boxes have `Origin = Anchor.TopLeft` by default. Should they have `Origin = Anchor.TopCentre`, the last one should be drawn too. This would probably lead to a better result, I'll try it and see.